### PR TITLE
Create ml-problem-formulation-envisioning.md

### PR DESCRIPTION
**Pull request by Chris-hughes10**
Added a page on problem formulation and envisioning to the ml-fundamentals folder. Specifically, this covers:
- General guidance on problem formulation
- Checklists to be used during envisioning sessions
- An example of breaking down a problem and forming hypothesis statements

### CODE REVIEW

**Comment by omri374** [comment]
Great content @Chris-hughes10! Awesome addition to the ML fundamentals.

A few minor suggestions, most if not all are subjective:
1. Consider having a "Goals" section on what should be the output of the envisioning phase.

2. Consider defining the boundary between envisioning and feasibility study. To me, some of the topics mentioned here felt more natural to be discussed in the feasibility study. For example, the hypotheses discussion is something we might start during envisioning and continue during feasibility study, but the boundary for which hypothesis goes to which part isn't clear. Are there specific hypotheses we would make during envisioning and prior to the feasibility study? 
3. The recommendation example is great. However, I found the last part around specific features or data sources to be a bit too detailed (both for this document and for the envisioning phase of the hypothetical project, where we are probably still missing some clarity on what data we really have as we haven't actively started digging into data). Again, it's just my view on this, not necessarily the right one.
4. Is problem formulation a subtask of the envisioning process or do you see them as two separate tasks? 
5. One approach I learned about and found interesting around envisioning, is to use a template for problem formulation (taken from [this blog post](https://blogs.intuit.com/blog/2020/06/23/lessons-learned-leading-ai-teams/)). The authors use this template:

    ```
    “<Customer> is trying to gain this <benefit> but is unable to because of <problem>. 
    We will help deliver this <benefit> by <means/solution>, 
    which will lead to <improvement from value to value>, and delivered by <delivery date>.“
    ```

    Could be interesting to integrate something like that here too.

6. Consider discussing success criteria for the envisioning part, or warning signs for an non-optimal envisioning process that might risk the project if not answered.
7. Perhaps rename "Envisioning checklist" to "Envisioning steps" or similar. A checklist (to me) sounds more focused on verifying you have everything needed (yes/no questions).

This is another relevant resource for this part of the ML project: https://docs.microsoft.com/en-us/azure/machine-learning/team-data-science-process/lifecycle-business-understanding

**Comment by Chris-hughes10** [comment]
Hi @omri374, thanks for the review. I will address your suggestions in turn:

`Consider having a "Goals" section on what should be the output of the envisioning phase.`
We can definitely do this.

`Consider defining the boundary between envisioning and feasibility study. To me, some of the topics mentioned here felt more natural to be discussed in the feasibility study. For example, the hypotheses discussion is something we might start during envisioning and continue during feasibility study, but the boundary for which hypothesis goes to which part isn't clear. Are there specific hypotheses we would make during envisioning and prior to the feasibility study?`
For me, everything discussed here is relevant to envisioning and not the feasibility study. Personally, I like to define hypothesis statements before starting a feasibility study so that there is a clear direction to explore. Of course, this can lead to further hypothesis statements being defined based on these during the study, but I believe that at least one hypothesis statement to guide the feasibility study should be defined during envisioning; personally I would try to define as many as possible.

`The recommendation example is great. However, I found the last part around specific features or data sources to be a bit too detailed (both for this document and for the envisioning phase of the hypothetical project, where we are probably still missing some clarity on what data we really have as we haven't actively started digging into data). Again, it's just my view on this, not necessarily the right one.`
I understand your point, but I disagree with this one. Even without seeing the data, I think it can be valuable to start to think about what data would be potentially useful to solve the problem. If it turns out that some of this data doesn't exist, a recommendation can still be provided to consider collecting it in future. In terms of the level of detail in the example, for me, I tend get a lot more value from specific, concrete examples when reading about new processes so I really wanted to illustrate the process from vague objective to specific hypothesis statements.

`Is problem formulation a subtask of the envisioning process or do you see them as two separate tasks?`
I see problem formulation as the overall goal of the envisioning process. Adding a goals section, as mentioned above, will hopefully make this clearer.

`One approach I learned about and found interesting around envisioning, is to use a template for problem formulation (taken from this blog post) [...]`
This looks interesting, and I am open to integrating this, but is not something that I have personally used so I would welcome some contributions on this.

`Consider discussing success criteria for the envisioning part, or warning signs for an non-optimal envisioning process that might risk the project if not answered.`
I think that the success criteria seems very similar to the proposed 'goals' section; I would prefer to include one or the other to stop the doc getting huge. I think that the warning signs section would be a great addition! However, it is standalone from the content that is already here, so I would recommend focusing on getting the current content merged as a starting point, and this can be added in a separate PR.

`Perhaps rename "Envisioning checklist" to "Envisioning steps" or similar. A checklist (to me) sounds more focused on verifying you have everything needed (yes/no questions).`
Happy to do this, I deliberated over using checklist for the same reason.

Let me know what you think, and I will go ahead and make the changes once another review has been given.

**Comment by omri374** [comment]
Hi @Chris-hughes10, thanks for the detailed answers! Approving, feel free to keep as is or change based on the things you suggested on rephrasing.

**Review by omri374 — APPROVED** [review]

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+* What is the problem?
+* Why does the problem need to be solved?
+* Does this problem require a machine learning solution?
+* How would a potential solution be used?
+
+However, establishing this understanding can prove difficult, especially for those unfamiliar with the problem domain. To ease this process, we can approach problems in a structured way by taking the following steps:  
+
+* Identify a measurable problem and define this in business terms, such that there is a clear objective and a good understanding of the input elements that can be controlled which may affect this objective; here, it is best to be as specific as possible.
+* Decide how the performance of a solution should be measured, such that it is aligned with the business objective, and the data needed to evaluate the effectiveness of a solution. Note that the data required to evaluate a solution's impact on the overall objective is not necessarily the same as that which may be required to create a solution.
```
Comment (biancafbmd) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:19]:
```suggestion
* Decide how the performance of a solution should be measured and identify if it can be measured within the restrictions of this problem, such that it is aligned with the business objective, and the data needed to evaluate the effectiveness of a solution. Note that the data required to evaluate a solution's impact on the overall objective is not necessarily the same as that which may be required to create a solution.
```

Comment (biancafbmd) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:19]: Maybe not the best suggestion, but wanted to clarify that sometimes the performance cannot be measured with the data/systems at your disposal.

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+* Based on the available data, define specific hypothesis statements - which can be proved
+or disproved - to guide the exploration of the data science team. Where possible, each hypothesis statement should have a clearly defined success criteria (e.g., *with an accuracy of over 60%*), however, this is not always possible - especially for projects where no solution to the problem currently exists. In these cases, the measure of success could be based on a subject matter expert verifying that the results meet their expectations.
+* Document all the above information, to ensure alignment between stakeholders and establish a clear understanding of the problem to be solved. Try to ensure that as much relevant domain knowledge is captured as possible, and that the features present in available data - and the way that the data was collected - are clearly explained, such that they can be understood by a non-subject matter expert.
+
+Once an understanding of the problem domain has been established, it may be necessary to break down the overall problem into smaller, meaningful chunks of work to maintain team focus and ensure a realistic project scope within the given timeframe.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:26]:
```suggestion
Once an understanding of the problem domain has been established, it may be necessary to break down the overall problem into smaller, meaningful chunks of work to maintain team focus and ensure a realistic project scope within the given time frame.
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+We would then be able to use this data to explore:
+
+* a method of measuring similarity between items
+* a method of measuring similarity between customers
+* a method of measuring how complementary items are relative to one another
+
+which can be used to create and rank recommendations. Depending on the project scope, and available data, one or more of these areas could be selected to create hypotheses to be explored by the data science team. Some examples of such hypothesis statements could be:
+
+* From the descriptions of each item, we can determine a measure of similarity between different items to a degree of accuracy which is specified by a stylist.
+* Based on the behaviour of customers with similar purchasing histories, we are able to predict certain items that a customer is likely to purchase; with a certainty which is greater than random choice.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:106]:
```suggestion
* Based on the behavior of customers with similar purchasing histories, we are able to predict certain items that a customer is likely to purchase; with a certainty which is greater than random choice.
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+2. How will the solution be used?
+3. What are the current solutions/workarounds (if any)?
+4. How should performance be measured?
+5. Is the performance measure aligned with the business objective?
+6. What would be the minimum performance needed to reach the business objective?
+7. Are there any known constraints around non-functional requirements that would have to be taken into account? (e.g., computation times)
+8. Frame this problem (supervised/unsupervised, online/offline, etc.)
+9. Is human expertise available?
+10. How would you solve the problem manually?
+11. What work has been done in this area so far?
```
Comment (biancafbmd) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:44]: Is 11 not super similar to 3? maybe the two can be merged

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+Once an understanding of the problem domain has been established, it may be necessary to break down the overall problem into smaller, meaningful chunks of work to maintain team focus and ensure a realistic project scope within the given timeframe.
+
+## Envisioning Checklist
+
+During envisioning sessions, the following may prove useful for guiding the discussion.
+
+### Problem Framing
+
+1. Define the objective in business terms.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:34]: You don't have to change it here - but ideally, we should set all numbers to 1. so that the markdown renderer auto numbers (so we don't end up with stale numbering if someone changes the list)  

For this particular list it seems that the items are not necessarily ordered, should this be an unordered list?

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+5. Is the performance measure aligned with the business objective?
+6. What would be the minimum performance needed to reach the business objective?
+7. Are there any known constraints around non-functional requirements that would have to be taken into account? (e.g., computation times)
+8. Frame this problem (supervised/unsupervised, online/offline, etc.)
+9. Is human expertise available?
+10. How would you solve the problem manually?
+11. What work has been done in this area so far?
+12. Are there any restrictions on the type of approaches which can be used? (e.g., does the solution need to be completely explainable?)
+13. List the assumptions you or others have made so far. Verify these assumptions if possible.
+14. Define some hypothesis statements to be explored.
```
Comment (biancafbmd) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:47]:
```suggestion
14. Define some initial hypothesis statements to be explored.
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+* From the descriptions of each item, we can determine a measure of similarity between different items to a degree of accuracy which is specified by a stylist.
+* Based on the behaviour of customers with similar purchasing histories, we are able to predict certain items that a customer is likely to purchase; with a certainty which is greater than random choice.
+* Using sets of items which have previously been sold together, we can formulate rules around the features which determine whether items are complementary or not which can be verified by a stylist.
+
+## References
+
+Many of the ideas presented here - and much more - were inspired by, and can be found in the following resources; all of which are highly recommended.
+
+* Fast.ai's Data project checklist: <https://www.fast.ai/2020/01/07/data-questionnaire/>
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:113]:
```suggestion
* [Fast.ai's Data project checklist](https://www.fast.ai/2020/01/07/data-questionnaire/)
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+* Aurélien Géron's Machine learning project checklist: <https://github.com/ageron/handson-ml/blob/master/ml-project-checklist.md>
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:114]:
```suggestion
* [Aurélien Géron's Machine learning project checklist](https://github.com/ageron/handson-ml/blob/master/ml-project-checklist.md)
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+* Designing great data products. Jeremy Howard, Margit Zwemer and Mike Loukides: <https://www.oreilly.com/radar/drivetrain-approach-data-products/>
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:115]:
```suggestion
* [Designing great data products. Jeremy Howard, Margit Zwemer and Mike Loukides](https://www.oreilly.com/radar/drivetrain-approach-data-products/)
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+6. What would be the minimum performance needed to reach the business objective?
+7. Are there any known constraints around non-functional requirements that would have to be taken into account? (e.g., computation times)
+8. Frame this problem (supervised/unsupervised, online/offline, etc.)
+9. Is human expertise available?
+10. How would you solve the problem manually?
+11. What work has been done in this area so far?
+12. Are there any restrictions on the type of approaches which can be used? (e.g., does the solution need to be completely explainable?)
+13. List the assumptions you or others have made so far. Verify these assumptions if possible.
+14. Define some hypothesis statements to be explored.
+15. Highlight and discuss any responsible AI concerns if appropriate.
```
Comment (biancafbmd) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:48]: Other things that could be added here:
- How will this solution be maintained and monitored? 
- Does this solution need to fit within an existing system?

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
+# Envisioning & Problem formulation
+
+Before beginning a data science investigation, it is necessary to have first defined problem statement which the data science team can explore; the manner of this definition can have a significant influence on whether the project is likely to be successful.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:3]:
```suggestion
Before beginning a data science investigation, we need to define a problem statement which the data science team can explore; this problem statement can have a significant influence on whether the project is likely to be successful.  
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+Generally, before defining a project scope for a data science investigation, we must first understand the problem domain:
+
+* What is the problem?
+* Why does the problem need to be solved?
+* Does this problem require a machine learning solution?
+* How would a potential solution be used?
+
+However, establishing this understanding can prove difficult, especially for those unfamiliar with the problem domain. To ease this process, we can approach problems in a structured way by taking the following steps:  
+
+* Identify a measurable problem and define this in business terms, such that there is a clear objective and a good understanding of the input elements that can be controlled which may affect this objective; here, it is best to be as specific as possible.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:18]:
```suggestion
* Identify a measurable problem and define this in business terms. The objective should be clear, and we should have a good understanding of the inputs and how they affect the objective. Be as specific as possible.
```

Comment (TessFerrandez) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:18]: I found it hard to follow the sentence so I tried to reformulate it, but am unsure if I captured what you wanted to say.

**Review by biancafbmd — CHANGES_REQUESTED** [review]
Great addition @Chris-hughes10 . A few suggestions/comments below:

* Could you link to this page from the main ML Fundamentals README? (there is an Envisioning reference under the Project Phases)
* A "goals" section would be a great addition to this imo. 
* I would agree with the discussion about re. renaming "Envisioning checklist" to "Envisioning steps"/"Envisioning guidelines".
* The items in the "Envisioning checklist" are numbered. Does this mean the order is expected to be sequential? If not, it might be better to use bullet points rather than a numbered list. 
* I think the problem formulation template is a great suggestion. I will add it to the backlog and start iterating on an option for that for a future PR.

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 1 above)
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:19]: I am finding it a little bit hard to follow this sentence as well, can it be formulated a bit more concisely? How is the performance aligned to the data needed. and what does it mean that the data required to evaluate a solutions impact on the overall objective...

I have made a suggestion, but again, not sure if I got it correctly.

Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:19]:
```suggestion
* Decide how the performance of a solution should be measured. Make sure it aligns with the business objective and that you have the data required to evaluate the solution. Note that the data required to evaluate a solution may differ from the data needed to create a solution.
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+* Decide how the performance of a solution should be measured, such that it is aligned with the business objective, and the data needed to evaluate the effectiveness of a solution. Note that the data required to evaluate a solution's impact on the overall objective is not necessarily the same as that which may be required to create a solution.
+* Thinking about the solution as a black box, detail the function that a solution to this problem should perform to fulfil the objective and verify that the relevant data is available to solve the problem.
+  * One way of approaching this is by thinking about how a subject matter expert could solve the problem manually and the data that would be required; if a human subject matter expert is unable to use the available data to solve the problem, this is indicative that additional information is required and/or more data needs to be collected.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:21]:
```suggestion
  * One way of approaching this is by thinking about how a subject matter expert could solve the problem manually and the data that would be required; if a human subject matter expert is unable to solve the problem given the available data, this is indicative that additional information is required and/or more data needs to be collected.
```

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+* Based on the available data, define specific hypothesis statements - which can be proved
+or disproved - to guide the exploration of the data science team. Where possible, each hypothesis statement should have a clearly defined success criteria (e.g., *with an accuracy of over 60%*), however, this is not always possible - especially for projects where no solution to the problem currently exists. In these cases, the measure of success could be based on a subject matter expert verifying that the results meet their expectations.
+* Document all the above information, to ensure alignment between stakeholders and establish a clear understanding of the problem to be solved. Try to ensure that as much relevant domain knowledge is captured as possible, and that the features present in available data - and the way that the data was collected - are clearly explained, such that they can be understood by a non-subject matter expert.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:24]: Are there some good templates or tools to document this? maybe a definition of done, or some ADR like template?

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+10. How would you solve the problem manually?
+11. What work has been done in this area so far?
+12. Are there any restrictions on the type of approaches which can be used? (e.g., does the solution need to be completely explainable?)
+13. List the assumptions you or others have made so far. Verify these assumptions if possible.
+14. Define some hypothesis statements to be explored.
+15. Highlight and discuss any responsible AI concerns if appropriate.
+
+### Data Exploration
+
+1. Understand and document the features, location, and availability of the data.
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:52]: Same comments as above

File: ml-fundamentals/ml-problem-formulation-envisioning.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+3. How does the organization decide when to collect additional data or purchase external data? Are there any examples of this?
+4. What data has been used so far to analyse recent data-driven projects? What has been found to be most useful? What was not useful? How was this judged?
+5. What additional internal data may provide insights useful for data-driven decision making for proposed projects? What external data could be useful?
+6. What are the possible constraints or challenges in accessing or incorporating this data?
+7. How was the data collected? Are there any obvious biases due to how the data was collected?
+8. What changes to data collection, coding, integration, etc has occurred in the last 2 years that may impact the interpretation or availability of the collected data
+
+### Workflow
+
+1. What data science skills are currently in the organization?
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-problem-formulation-envisioning.md:63]:
```suggestion
1. What data science skills exist in the organization?
```

**Review by TessFerrandez — CHANGES_REQUESTED** [review]
Great content - had some suggestions for rewording a few sections to increase readability, and some formatting nits (like the markdown links) - to follow the markdown guidelines for the repo, but overall a great doc.

**Comment by Chris-hughes10** [comment]
Comment to remind myself: Improve the referencing before merge

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 7 above)
Comment (Chris-hughes10) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:113]: The `<` around links were added based on the linter that was recommended in the contribution guidelines, will it complain if I remove these?

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 3 above)
Comment (Chris-hughes10) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:106]: Are we going U.S or UK? ;)

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 7 above)
Comment (TessFerrandez) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:113]: You got that warning because you had bare urls - but in general we try to use markdown linking - I.e with []() and link the words rather than the urls in the playbook - the suggestion should work

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 3 above)
Comment (TessFerrandez) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:106]: Fair point :) I go through most PRs and have tried to standardize on US English for a cohesive feeling - but having said that. I won’t reject if you refuse to change this

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 14 above)
Comment (Chris-hughes10) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:33]: Not that I am aware of, that could be a good area of future work

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 5 above)
Comment (Chris-hughes10) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:43]: I think that there is a rough ordering, so I have left the numbers.

**Comment by Chris-hughes10** [comment]
cc: @biancafbmd @TessFerrandez I have updated the doc based on the changes suggested.

**Comment by Chris-hughes10** [comment]
With regards to the failing check. It looks like a link failing that is unrelated to this work.

File: ml-fundamentals/ml-problem-formulation-envisioning.md (hunk 5 above)
Comment (TessFerrandez) [reply ml-fundamentals/ml-problem-formulation-envisioning.md:43]: Yeah, I meant as in if you set all the numbers to 1 markdown will automatically give the right numbers, that way if someone removes an item from the list later, it will re-order correctly

**Review by TessFerrandez — COMMENTED** [review]
> With regards to the failing check. It looks like a link failing that is unrelated to this work.

Yes, its fixed in the main branch - so you need to merge the latest changes from main into your fork before you will be able to merge in your PR. With that link failing your merge is blocked

**Comment by Chris-hughes10** [comment]
Even after merging, I am still failing, but it seems to be different links. Have I gone wrong somewhere?

**Review by biancafbmd — APPROVED** [review]

**Comment by TessFerrandez** [comment]
> Even after merging, I am still failing, but it seems to be different links. Have I gone wrong somewhere?
I checked all failing links - and they seem ok.  Sometimes that link checker can be a bit finicky, it will report links as dead if they don't respond quickly enough... I reran 2 times and now its passing, so let me know when you're done with all the requested changes and I'll approve too so you can merge.

**Review by TessFerrandez — APPROVED** [review]
